src/main/java/fi/fmi/avi/archiver/file/FileParser.java

```diff
@@ -0,0 +1,71 @@
+import org.springframework.integration.annotation.ServiceActivator;
+import org.springframework.messaging.MessageHeaders;
+
+import javax.annotation.Nullable;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static java.util.Objects.requireNonNull;
+
+public class FileParser {
+
+    private final AviMessageConverter aviMessageConverter;
+
+    public FileParser(final AviMessageConverter aviMessageConverter) {
+        this.aviMessageConverter = requireNonNull(aviMessageConverter, "aviMessageConverter");
+    }
+
+    @ServiceActivator
+    public List<FileAviationMessage> parse(final String content, final MessageHeaders headers) {
+        final FilenamePattern pattern = (FilenamePattern) headers.get(MessageFileMonitorInitializer.MESSAGE_FILE_PATTERN);
+        final Instant fileModified = (Instant) headers.get(MessageFileMonitorInitializer.FILE_MODIFIED);
+        final String productIdentifier = (String) headers.get(MessageFileMonitorInitializer.PRODUCT_IDENTIFIER);
+        return parse(productIdentifier, pattern, content.trim(), fileModified);
+    }
+
+    public List<FileAviationMessage> parse(final String productIdentifier, final FilenamePattern filenamePattern, final String content,
+                                           @Nullable final Instant fileModified) {
+        final ConversionResult<GenericMeteorologicalBulletin> bulletinConversion = aviMessageConverter.convertMessage(content,
+                TACConverter.TAC_TO_GENERIC_BULLETIN_POJO);
+        if (!bulletinConversion.getConvertedMessage().isPresent()) {
+            // Unable to get any messages out of the file content
+            return Collections.emptyList();
+        }
+
+        final GenericMeteorologicalBulletin bulletin = bulletinConversion.getConvertedMessage().get();
+        final FileBulletinHeading gtsBulletinHeading = FileBulletinHeading.builder()
+                .setBulletinHeading(bulletin.getHeading())
+                .setBulletinHeadingString(BulletinHeadingEncoder.encode(bulletin.getHeading(), MessageFormat.TEXT, null))
+                .build();
+        final FileMetadata fileMetadata = FileMetadata.builder()
+                .setFilenamePattern(filenamePattern)//
+                .setProductIdentifier(productIdentifier)//
+                .setFileModified(fileModified)
+                .build();
+
+        return bulletin.getMessages().stream()//
+                .map(message -> { //
+                    if (!message.getMessageType().isPresent()) {
+                        throw new IllegalStateException("Message type cannot be determined");
+                    }
+
+                    return FileAviationMessage.builder()//
+                            .setGtsBulletinHeading(gtsBulletinHeading)
+                            .setFileMetadata(fileMetadata)
+                            .setMessage(message)
+                            .build();
```

Comment:
It doesn't even though it looks like it should. The tests do run through this part. Not sure what's going on there.

Reply:
I looked into this: it seems to be FreeBuilder magic. As both `InputAviationMessage` and `InputBulletinHeading` are FreeBuilder classes, the `InputAviationMessage` constructor builds an empty `InputBulletinHeading` value if not set. An excerpt from generated value class constructor:

```
if (builder.collectIdentifier == null) {
  this.collectIdentifier = InputBulletinHeading.builder().build();
```

This will fail if `InputBulletinHeading` class ever changes to hold a mandatory property without a default value. But such change requires a change in code under inspection anyway. As this code are covered by tests that will fail on such change, I see no need for any changes now.